### DISCUSSION TITLE
Official support for looking up resources by ID

### DISCUSSION THREAD

**Discussion by aterweele** [community/community#24515] [Apps, API and Webhooks]
<p>Hello! Is there a way to look up users, repositories, and organizations by their unchanging ID? There are the endpoints mentioned in <a href="https://stackoverflow.com/a/30579888" rel="nofollow noopener">https://stackoverflow.com/a/30579888</a>, but none of them are attested in API documentation. They do seem to appear in the official Ruby client, however:</p>
<ul>
<li><a href="https://github.com/octokit/octokit.rb/blob/a201a5ca366d407122f130cd2507a1b716a0a89d/lib/octokit/user.rb#L13" rel="nofollow noopener">https://github.com/octokit/octokit.rb/blob/a201a5ca366d407122f130cd2507a1b716a0a89d/lib/octokit/user.rb#L13</a></li>
<li>…/repository.rb#L68</li>
<li>…/organization.rb#L13</li>
</ul>
<p>(edited down because as a new user I can only post two links)</p>
<p>Will those endpoints be made official and documented?</p>
<p>Alternatively, is there a way to look up these resources by ID in the GraphQL API?</p>

**Comment by francisfuzz** [comment, marked as answer]
<p>:wave: <a class="mention" href="/aterweele">@aterweele</a>, welcome to the GitHub Support Community. We appreciate you asking these questions here!</p>
<aside class="quote no-group" data-username="aterweele" data-post="1" data-topic="119703">
<div class="title">
<div class="quote-controls"></div>
 aterweele:</div>
<blockquote>
<p>Will those endpoints be made official and documented?</p>
</blockquote>
</aside>
<p>We can’t speak to specific timelines for if or when endpoints will be made official and documented. That said, we checked internally with our team and have an open issue discussing this. For now, we recommend following <a href="http://github.blog/">http://github.blog/</a> and <a href="https://developer.github.com/changes" class="inline-onebox">GitHub API Changes | GitHub Developer Guide</a> for the latest updates to <a href="http://GitHub.com">GitHub.com</a> and our API respectively.</p>
<p>If you have any other notes to share about your use case, would you mind <a href="https://support.github.com/contact/feedback">submitting this through our official product feedback form</a> so that our product team can track your request? That’s the best place to share requests like these in consideration for future iterations of GitHub features.</p>
<aside class="quote no-group">
<blockquote>
<p>Alternatively, is there a way to look up these resources by ID in the GraphQL API?</p>
</blockquote>
</aside>
<p>Yes, but I think some additional context is warranted.</p>
<p>As you observed, the GitHub REST API can resolve a repository by its database ID. It can also resolve an organization or a user account by its database ID.</p>
<p>The GitHub GraphQL API’s <a href="https://developer.github.com/v4/query">query root</a> exposes the following fields:</p>
<ul>
<li>a <a href="https://developer.github.com/v4/query/#node"><code>node</code> field</a> to fetch an object given its node ID.</li>
<li>a <a href="https://developer.github.com/v4/query/#repository"><code>repository</code> field</a> to lookup a given repository by the owner and repository name, but no arguments exist for lookup by database ID.</li>
<li>a <a href="https://developer.github.com/v4/query/#user"><code>user</code> field</a> to lookup a given repository by the owner and repository name, but no arguments exist for lookup by database ID.</li>
</ul>
<p>A node ID and database ID effectively point to the same resource, but their values are formatted differently.</p>
<p>Let’s use <a href="https://github.com/github/linguist"><code>github/linguist</code></a> as an example. We can get this repository via the GitHub REST API:</p>
<pre><code class="lang-auto">curl -v -H "Authorization: token $TOKEN" "https://api.github.com/repos/github/linguist"
</code></pre>
<p>The response exposes its <code>id</code> and its <code>node_id</code>:</p>
<pre><code class="lang-auto">{
  "id": 1725199,
  "node_id": "MDEwOlJlcG9zaXRvcnkxNzI1MTk5",
  // ... truncated for brevity
}
</code></pre>
<p>Here’s a GraphQL query for fetching it by its node ID:</p>
<pre><code class="lang-auto">query {
  node(id:"MDEwOlJlcG9zaXRvcnkxNzI1MTk5") {
    ... on Repository {
      nameWithOwner
    }
  }
}
</code></pre>
<p>That <code>... on Repository</code> is an <a href="https://graphql.org/learn/queries/#inline-fragments">inline fragment</a> since a <a href="https://developer.github.com/v4/interface/node/">Node</a> is implemented by many other types that communicates that if the node ID happens to match that <code>Repository</code> object, expand it and list its <code>nameWithOwner</code>. It’s possible that your query may target more than just repositories and you may want to expand certain fields differently. I hope this helps!</p>
